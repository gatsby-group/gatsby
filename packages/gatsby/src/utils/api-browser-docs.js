/**
 * Called when the Gatsby browser runtime first starts.
 * @example
 * exports.onClientEntry = () => {
 *   console.log("We've started!")
 *   callAnalyticsAPI()
 * }
 */
exports.onClientEntry = true

/**
 * Called when gatsby is fetching resources for a page longer than 1 second.
 * @example
 * exports.onFetchingResources = () => {
 *   console.log("We can show loading indicator")
 * }
 */
exports.onFetchingResources = true

/**
 * Called when gatsby has fetched all resources for page
 * @example
 * exports.onFetchedResources = () => {
 *   console.log("We can hide loading indicator")
 * }
 */
exports.onFetchedResources = true

/**
 * Called when the initial (but not subsequent) render of Gatsby App is done on the client.
 * @example
 * exports.onInitialClientRender = () => {
 *   console.log("ReactDOM.render has executed")
 * }
 */
exports.onInitialClientRender = true

/**
 * Called when the user changes routes
 * @param {object} $0
 * @param {object} $0.location A location object
 * @param {object} $0.action The "action" that caused the route change
 * @example
 * exports.onRouteUpdate = ({ location }) => {
 *   console.log('new pathname', location.pathname)
 * }
 */
exports.onRouteUpdate = true

/**
 * Allow a plugin to decide if the "scroll" should update or
 * not on a route change.
 * @param {object} $0
 * @param {object} $0.prevRouterProps The previous state of the router before the route change.
 * @param {object} $0.pathname The new pathname
 */
exports.shouldUpdateScroll = true

/**
 * Allow a plugin to register a Service Worker. Should be a function that returns true.
 * @example
 * exports.registerServiceWorker = () => true
 */
exports.registerServiceWorker = true

/**
 * Allow a plugin to replace the router component e.g. to use a custom history version.
 * @param {object} $0
 * @param {object} $0.history The history instance to use in the replacement router instance
 */
exports.replaceRouterComponent = true

/**
 * Allow a plugin to replace the page component renderer. This api runner can be used to
 * implement page transitions. See https://github.com/gatsbyjs/gatsby/tree/master/examples/using-page-transitions for an example of this.
 * @param {object} $0
 * @param {object} $0.props The props of the page.
 * @param {object} $0.loader The gatsby loader.
 */
exports.replaceComponentRenderer = true

/**
 * Allow a plugin to replace the history object.
 */
exports.replaceHistory = true

/**
 * Allow a plugin to wrap the root component.
 * @param {object} $0
 * @param {object} $0.Root The "Root" component built by Gatsby.
 */
exports.wrapRootComponent = true
