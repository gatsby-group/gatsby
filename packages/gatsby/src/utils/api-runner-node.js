const Promise = require(`bluebird`)
const glob = require(`glob`)
const _ = require(`lodash`)
const chalk = require(`chalk`)

const tracer = require(`opentracing`).globalTracer()
const reporter = require(`gatsby-cli/lib/reporter`)
const getCache = require(`./get-cache`)
const apiList = require(`./api-node-docs`)
const createNodeId = require(`./create-node-id`)
const createContentDigest = require(`./create-content-digest`)
const { getNonGatsbyCodeFrame } = require(`./stack-trace-utils`)

// Bind action creators per plugin so we can auto-add
// metadata to actions they create.
const boundPluginActionCreators = {}
const doubleBind = (boundActionCreators, api, plugin, actionOptions) => {
  const { traceId } = actionOptions
  if (boundPluginActionCreators[plugin.name + api + traceId]) {
    return boundPluginActionCreators[plugin.name + api + traceId]
  } else {
    const keys = Object.keys(boundActionCreators)
    const doubleBoundActionCreators = {}
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      const boundActionCreator = boundActionCreators[key]
      if (typeof boundActionCreator === `function`) {
        doubleBoundActionCreators[key] = (...args) => {
          // Let action callers override who the plugin is. Shouldn't be
          // used that often.
          if (args.length === 1) {
            return boundActionCreator(args[0], plugin, actionOptions)
          } else if (args.length === 2) {
            return boundActionCreator(args[0], args[1], actionOptions)
          }
        }
      }
    }
    boundPluginActionCreators[
      plugin.name + api + traceId
    ] = doubleBoundActionCreators
    return doubleBoundActionCreators
  }
}

const initAPICallTracing = parentSpan => {
  const startSpan = (spanName, spanArgs = {}) => {
    const defaultSpanArgs = { childOf: parentSpan }

    return tracer.startSpan(spanName, _.merge(defaultSpanArgs, spanArgs))
  }

  return {
    tracer,
    parentSpan,
    startSpan,
  }
}

const runAPI = (plugin, api, args) => {
  const gatsbyNode = require(`${plugin.resolve}/gatsby-node`)
  if (gatsbyNode[api]) {
    const parentSpan = args && args.parentSpan
    const spanOptions = parentSpan ? { childOf: parentSpan } : {}
    const pluginSpan = tracer.startSpan(`run-plugin`, spanOptions)

    pluginSpan.setTag(`api`, api)
    pluginSpan.setTag(`plugin`, plugin.name)

    let pathPrefix = ``
    const { store, emitter } = require(`../redux`)
    const {
      loadNodeContent,
      getNodes,
      getNode,
      getNodesByType,
      hasNodeChanged,
      getNodeAndSavePathDependency,
    } = require(`../db/nodes`)
    const { boundActionCreators } = require(`../redux/actions`)

    const doubleBoundActionCreators = doubleBind(
      boundActionCreators,
      api,
      plugin,
      { ...args, parentSpan: pluginSpan }
    )

    if (store.getState().program.prefixPaths) {
      pathPrefix = store.getState().config.pathPrefix
    }

    const namespacedCreateNodeId = id => createNodeId(id, plugin.name)

    const tracing = initAPICallTracing(pluginSpan)

    const cache = getCache(plugin.name)

    // Ideally this would be more abstracted and applied to more situations, but right now
    // this can be potentially breaking so targeting `createPages` API and `createPage` action
    let apiFinished = false
    if (api === `createPages`) {
      let alreadyDisplayed = false
      const createPageAction = doubleBoundActionCreators.createPage
      doubleBoundActionCreators.createPage = async (...args) => {
        if (apiFinished && !alreadyDisplayed) {
          const warning = [
            reporter.stripIndent(`
              Action ${chalk.bold(
                `createPage`
              )} was called outside of its expected asynchronous lifecycle ${chalk.bold(
              `createPages`
            )} in ${chalk.bold(plugin.name)}.
              Ensure that you return a Promise from ${chalk.bold(
                `createPages`
              )} and are awaiting any asynchronous method invocations (like ${chalk.bold(
              `graphql`
            )} or http requests).
              For more info and debugging tips: see ${chalk.bold(
                `https://gatsby.app/sync-actions`
              )}
            `),
          ]

          const possiblyCodeFrame = getNonGatsbyCodeFrame()
          if (possiblyCodeFrame) {
            warning.push(possiblyCodeFrame)
          }

          reporter.warn(warning.join(`\n\n`))
          alreadyDisplayed = true
        }
        return createPageAction(...args)
      }
    }

    const apiCallArgs = [
      {
        ...args,
        pathPrefix,
        boundActionCreators: doubleBoundActionCreators,
        actions: doubleBoundActionCreators,
        loadNodeContent,
        store,
        emitter,
        getCache,
        getNodes,
        getNode,
        getNodesByType,
        hasNodeChanged,
        reporter,
        getNodeAndSavePathDependency,
        cache,
        createNodeId: namespacedCreateNodeId,
        createContentDigest,
        tracing,
      },
      plugin.pluginOptions,
    ]

    // If the plugin is using a callback use that otherwise
    // expect a Promise to be returned.
    if (gatsbyNode[api].length === 3) {
      return Promise.fromCallback(callback => {
        const cb = (err, val) => {
          pluginSpan.finish()
          callback(err, val)
          apiFinished = true
        }
        gatsbyNode[api](...apiCallArgs, cb)
      })
    } else {
      const result = gatsbyNode[api](...apiCallArgs)
      pluginSpan.finish()
      return Promise.resolve(result).then(res => {
        apiFinished = true
        return res
      })
    }
  }

  return Promise.resolve()
}

let filteredPlugins
const hasAPIFile = plugin => glob.sync(`${plugin.resolve}/gatsby-node*`)[0]

module.exports = async (api, args = {}, pluginSource) => {
  const { parentSpan } = args
  const apiSpanArgs = parentSpan ? { childOf: parentSpan } : {}
  const apiSpan = tracer.startSpan(`run-api`, apiSpanArgs)

  apiSpan.setTag(`api`, api)
  _.forEach(args.traceTags, (value, key) => {
    apiSpan.setTag(key, value)
  })

  // Check that the API is documented.
  // "FAKE_API_CALL" is used when code needs to trigger something to happen
  // once the the API queue is empty. Ideally of course we'd have an API
  // (returning a promise) for that. But this works nicely in the meantime.
  if (!apiList[api] && api !== `FAKE_API_CALL`) {
    reporter.panic(`api: "${api}" is not a valid Gatsby api`)
  }

  const { store } = require(`../redux`)
  const plugins = store.getState().flattenedPlugins
  // Get the list of plugins that implement gatsby-node.
  if (!filteredPlugins) {
    filteredPlugins = plugins.filter(plugin => hasAPIFile(plugin))
  }

  // Break infinite loops.
  // Sometimes a plugin will implement an API and call an action which will
  // trigger the same API being called. `onCreatePage` is the only example
  // right now. In these cases, we should avoid calling the originating plugin
  // again.
  const noSourcePluginPlugins = pluginSource
    ? filteredPlugins.filter(p => p.name !== pluginSource)
    : filteredPlugins

  const results = []
  for (const plugin of noSourcePluginPlugins) {
    try {
      const result = await runAPI(plugin, api, { ...args, parentSpan: apiSpan })
      results.push(result)
    } catch (err) {
      const pluginName =
        plugin.name === `default-site-plugin`
          ? `gatsby-node.js`
          : `Plugin ${plugin.name}`
      reporter.panicOnBuild(`${pluginName} returned an error`, err)
    }
  }

  const { emitter } = require(`../redux`)
  emitter.emit(`API_RUNNING_QUEUE_EMPTY`)
  apiSpan.finish()

  return results.filter(result => !_.isEmpty(result))
}
