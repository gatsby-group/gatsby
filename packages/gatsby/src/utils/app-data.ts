import path from "path"

import { outputJson, pathExistsSync } from "fs-extra"

const APP_DATA_JSON = `app-data.json`

export const write = (publicDir: string, hash: string): Promise<void> =>
  outputJson(path.join(publicDir, `page-data`, APP_DATA_JSON), {
    webpackCompilationHash: hash,
  })

export const exists = (publicDir: string): boolean =>
  pathExistsSync(path.join(publicDir, `page-data`, APP_DATA_JSON))
