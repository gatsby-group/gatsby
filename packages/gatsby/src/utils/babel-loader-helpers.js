const path = require(`path`)
const _ = require(`lodash`)

const prepareOptions = babel => {
  let pluginBabelConfig = { test: { plugins: [], presets: [] } }
  if (process.env.NODE_ENV !== `test`) {
    const pluginBabelConfig = require(path.join(
      process.cwd(),
      `./.cache/babelState.json`
    ))
  }

  const stage = process.env.GATSBY_BUILD_STAGE || `test`

  // Required plugins/presets
  const requiredPlugins = [
    babel.createConfigItem(
      [require.resolve(`babel-plugin-remove-graphql-queries`)],
      {
        type: `plugin`,
      }
    ),
  ]
  const requiredPresets = []

  // Stage specific plugins to add
  if (stage === `build-html` || stage === `develop-html`) {
    requiredPlugins.push(
      babel.createConfigItem(
        [require.resolve(`babel-plugin-dynamic-import-node`)],
        {
          type: `plugin`,
        }
      )
    )
  }

  if (stage === `develop`) {
    requiredPlugins.push(
      babel.createConfigItem([require.resolve(`react-hot-loader/babel`)], {
        type: `plugin`,
      })
    )
  }

  // Fallback presets/plugins
  const fallbackPresets = []
  const fallbackPlugins = []

  let targets
  if (stage === `build-html`) {
    targets = {
      node: `current`,
    }
  } else {
    targets = {
      browsers: pluginBabelConfig.browserslist,
    }
  }

  fallbackPresets.push(
    babel.createConfigItem(
      [
        require.resolve(`@babel/preset-env`),
        {
          loose: true,
          modules: false,
          useBuiltIns: `usage`,
          targets,
        },
      ],
      {
        type: `preset`,
      }
    )
  )

  fallbackPresets.push(
    babel.createConfigItem(
      [
        require.resolve(`@babel/preset-react`),
        {
          useBuiltIns: true,
          pragma: `React.createElement`,
          development: stage === `develop`,
        },
      ],
      {
        type: `preset`,
      }
    )
  )

  fallbackPresets.push(
    babel.createConfigItem([require.resolve(`@babel/preset-flow`)], {
      type: `preset`,
    })
  )

  fallbackPlugins.push(
    babel.createConfigItem(
      [
        require.resolve(`@babel/plugin-proposal-class-properties`),
        {
          loose: true,
        },
      ],
      {
        type: `plugin`,
      }
    )
  )

  fallbackPlugins.push(
    babel.createConfigItem([require.resolve(`babel-plugin-macros`)], {
      type: `plugin`,
    })
  )

  fallbackPlugins.push(
    babel.createConfigItem(
      [require.resolve(`@babel/plugin-syntax-dynamic-import`)],
      {
        type: `plugin`,
      }
    )
  )

  fallbackPlugins.push(
    babel.createConfigItem(
      [
        require.resolve(`@babel/plugin-transform-runtime`),
        {
          helpers: true,
          regenerator: true,
          polyfill: false,
        },
      ],
      {
        type: `plugin`,
      }
    )
  )

  // Go through babel state and create config items for presets/plugins from.
  const reduxPlugins = []
  const reduxPresets = []
  console.log({ pluginBabelConfig, stage })
  console.log(pluginBabelConfig[stage])
  pluginBabelConfig[stage].plugins.forEach(plugin => {
    reduxPlugins.push(
      babel.createConfigItem([require.resolve(plugin.name), plugin.options], {
        type: `plugin`,
      })
    )
  })
  pluginBabelConfig[stage].presets.forEach(preset => {
    reduxPresets.push(
      babel.createConfigItem([require.resolve(preset.name), preset.options], {
        type: `preset`,
      })
    )
  })

  return [
    reduxPresets,
    reduxPlugins,
    requiredPresets,
    requiredPlugins,
    fallbackPlugins,
    fallbackPresets,
  ]
}

const mergeConfigItemOptions = ({ items, itemToMerge, type, babel }) => {
  const index = _.findIndex(
    items,
    i => i.file.resolved === itemToMerge.file.resolved
  )

  // If this exist, merge the options, otherwise, add it to the array
  if (index !== -1) {
    items[index] = babel.createConfigItem(
      [
        itemToMerge.file.resolved,
        _.merge({}, items[index].options, itemToMerge.options),
      ],
      {
        type,
      }
    )
  } else {
    items.push(itemToMerge)
  }

  return items
}

// Export helper functions for testing
exports.prepareOptions = prepareOptions
exports.mergeConfigItemOptions = mergeConfigItemOptions
