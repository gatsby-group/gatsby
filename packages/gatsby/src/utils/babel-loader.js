const babelLoader = require(`babel-loader`)

const {
  prepareOptions,
  getCustomOptions,
  mergeConfigItemOptions,
  addRequiredPresetOptions,
} = require(`./babel-loader-helpers`)
const { getBrowsersList } = require(`./browserslist`)

/**
 * Gatsby's custom loader for webpack & babel
 *
 * Gatsby allows sites to either use our Babel setup (the default)
 * or to add a .babelrc to take control.
 *
 * Our default setup is defined in the fallbackPlugins/fallbackPresets arrays
 * below.
 *
 * After using either the fallback or user supplied setup, we add on a handful
 * of required plugins and finally merge in any presets/plugins supplied
 * by Gatsby plugins.
 *
 * You can find documentation for the custom loader here: https://babeljs.io/docs/en/next/babel-core.html#loadpartialconfig
 */

const customOptionsCache = new Map()
const configCache = new Map()

module.exports = babelLoader.custom(babel => {
  return {
    // Passed the loader options.
    customOptions({
      stage = `test`,
      reactRuntime = `classic`,
      rootDir = process.cwd(),
      ...options
    }) {
      if (customOptionsCache.has(stage)) {
        return customOptionsCache.get(stage)
      }

      const toReturn = {
        custom: {
          stage,
          reactRuntime,
        },
        loader: {
          cacheIdentifier: JSON.stringify({
            browerslist: getBrowsersList(rootDir),
            babel: babel.version,
            gatsbyPreset: require(`babel-preset-gatsby/package.json`).version,
            env: babel.getEnv(),
          }),
          sourceType: `unambiguous`,
          ...getCustomOptions(stage),
          ...options,
        },
      }

      customOptionsCache.set(stage, toReturn)

      return toReturn
    },

    // Passed Babel's 'PartialConfig' object.
    config(partialConfig, { customOptions }) {
      let configCacheKey = customOptions.stage
      if (partialConfig.hasFilesystemConfig()) {
        // partialConfig.files is a Set that accumulates used config files (absolute paths)
        partialConfig.files.forEach(configFilePath => {
          configCacheKey += `_${configFilePath}`
        })
      }

      let { options } = partialConfig
      if (configCache.has(configCacheKey)) {
        return { ...options, ...configCache.get(configCacheKey) }
      }

      const [
        reduxPresets,
        reduxPlugins,
        requiredPresets,
        requiredPlugins,
        fallbackPresets,
      ] = prepareOptions(babel, customOptions)

      // If there is no filesystem babel config present, add our fallback
      // presets/plugins.
      if (!partialConfig.hasFilesystemConfig()) {
        options = {
          ...options,
          plugins: requiredPlugins,
          presets: [...fallbackPresets, ...requiredPresets],
        }
      } else {
        // With a babelrc present, only add our required plugins/presets
        options = {
          ...options,
          plugins: [...options.plugins, ...requiredPlugins],
          presets: [...options.presets, ...requiredPresets],
        }
        // User-defined preset likely contains `babel-preset-gatsby`
        // Make sure to pass required dynamic options (e.g. `stage` to it):
        addRequiredPresetOptions(babel, options.presets, customOptions)
      }

      // Merge in presets/plugins added from gatsby plugins.
      reduxPresets.forEach(preset => {
        options.presets = mergeConfigItemOptions({
          items: options.presets,
          itemToMerge: preset,
          type: `preset`,
          babel,
        })
      })

      reduxPlugins.forEach(plugin => {
        options.plugins = mergeConfigItemOptions({
          items: options.plugins,
          itemToMerge: plugin,
          type: `plugin`,
          babel,
        })
      })

      // cache just plugins and presets, because config also includes things like
      // filenames - this is mostly to not call `mergeConfigItemOptions` for each file
      // as that function call `babel.createConfigItem` and is quite expensive but also
      // skips quite a few nested loops on top of that
      configCache.set(configCacheKey, {
        plugins: options.plugins,
        presets: options.presets,
      })

      return options
    },
  }
})
