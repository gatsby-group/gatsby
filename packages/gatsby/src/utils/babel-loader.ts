import babelLoader from "babel-loader"
import {
  prepareOptions,
  getCustomOptions,
  mergeConfigItemOptions,
} from "./babel-loader-helpers"
import { getBrowsersList } from "./browserslist"
import { version } from "babel-preset-gatsby/package.json"

/**
 * Gatsby's custom loader for webpack & babel
 *
 * Gatsby allows sites to either use our Babel setup (the default)
 * or to add a .babelrc to take control.
 *
 * Our default setup is defined in the fallbackPlugins/fallbackPresets arrays
 * below.
 *
 * After using either the fallback or user supplied setup, we add on a handful
 * of required plugins and finally merge in any presets/plugins supplied
 * by Gatsby plugins.
 *
 * You can find documentation for the custom loader here: https://babeljs.io/docs/en/next/babel-core.html#loadpartialconfig
 */
export default babelLoader.custom((babel: Record<string, any>) => {
  const toReturn = {
    // Passed the loader options.
    customOptions({
      stage = `test`,
      reactRuntime = `classic`,
      rootDir = process.cwd(),
      ...options
    }: {
      stage?: string
      reactRuntime?: string
      rootDir?: string
    } = {}): { custom: Record<string, any>; loader: Record<string, any> } {
      return {
        custom: {
          stage,
          reactRuntime,
        },
        loader: {
          cacheIdentifier: JSON.stringify({
            browerslist: getBrowsersList(rootDir),
            babel: babel.version,
            gatsbyPreset: version,
            env: babel.getEnv(),
          }),
          sourceType: `unambiguous`,
          ...getCustomOptions(stage),
          ...options,
        },
      }
    },

    // Passed Babel's 'PartialConfig' object.
    config(
      partialConfig: Record<string, any>,
      { customOptions }: Record<string, any>
    ): Record<string, any> {
      let { options } = partialConfig
      const [
        reduxPresets,
        reduxPlugins,
        requiredPresets,
        requiredPlugins,
        fallbackPresets,
      ] = prepareOptions(babel, customOptions)

      // If there is no filesystem babel config present, add our fallback
      // presets/plugins.
      if (!partialConfig.hasFilesystemConfig()) {
        options = {
          ...options,
          plugins: requiredPlugins,
          presets: [...fallbackPresets, ...requiredPresets],
        }
      } else {
        // With a babelrc present, only add our required plugins/presets
        options = {
          ...options,
          plugins: [...options.plugins, ...requiredPlugins],
          presets: [...options.presets, ...requiredPresets],
        }
      }

      // Merge in presets/plugins added from gatsby plugins.
      reduxPresets.forEach((preset: Record<string, any>) => {
        options.presets = mergeConfigItemOptions({
          items: options.presets,
          itemToMerge: preset,
          type: `preset`,
          babel,
        })
      })

      reduxPlugins.forEach((plugin: Record<string, any>) => {
        options.plugins = mergeConfigItemOptions({
          items: options.plugins,
          itemToMerge: plugin,
          type: `plugin`,
          babel,
        })
      })

      return options
    },
  }

  return toReturn
})
