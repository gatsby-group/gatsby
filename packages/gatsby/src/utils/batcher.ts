export default class Batcher<Callback extends (...args: any[]) => any> {
  private queue: Parameters<Callback>[] = [];
  private callbacks: ((...args: Parameters<Callback>) => void)[] = []
  private bulkCallbacks: ((args: Parameters<Callback>[]) => void)[] = []

  constructor(private threshold: number) {}

  /** Add a call to the batcher */
  add(...args: Parameters<Callback>) {
    this.queue.push(args)

    if (this.queue.length >= this.threshold) {
      this.flush()
    }
  }

  /** Call all of our callbacks and clear out the queue */
  flush() {
    // call each callback for each item in the queue
    this.queue.forEach(args => 
      this.callbacks.forEach(callback => callback(...args))
    )

    // pass the entire queue to all bulk callbacks
    this.bulkCallbacks.forEach(callback => {
      callback(this.queue)
    })

    // clear out the queue
    this.queue = []
  }

  /** Sets up a callback for each batcher item */
  call(callback: (...args: Parameters<Callback>) => void) {
    this.callbacks.push(callback)
  }

  /** Sets up a bulk callback that takes the entire queue */
  bulkCall(callback: (args: Parameters<Callback>[]) => void) {
    this.bulkCallbacks.push(callback)
  }
}