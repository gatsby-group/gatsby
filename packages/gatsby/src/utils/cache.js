const fs = require(`fs-extra`)
const manager = require(`cache-manager`)
const fsStore = require(`cache-manager-fs`)
const path = require(`path`)

const MAX_CACHE_SIZE = 250

class Cache {
  constructor(folderName = `db`, store = fsStore) {
    this.folderName = folderName

    const caches = [
      { store: `memory`, max: MAX_CACHE_SIZE },
      {
        store,
        options: {
          path: this.directory,
        },
      },
    ].map(cache => manager.caching(cache))

    this.cache = manager.multiCaching(caches)
  }

  get directory() {
    return path.join(process.cwd(), `.cache/caches/${this.folderName}`)
  }

  init() {
    fs.ensureDirSync(this.directory)

    return this
  }

  get(key) {
    return this.cache.get(key)
  }

  set(key, value) {
    return this.cache.set(key, value)
  }
}

module.exports = Cache
