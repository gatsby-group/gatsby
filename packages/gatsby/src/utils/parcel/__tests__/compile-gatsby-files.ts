import { Parcel } from "@parcel/core"
import {
  constructParcel,
  compileGatsbyFiles,
  gatsbyFileRegex,
  COMPILED_CACHE_DIR,
  PARCEL_CACHE_DIR,
} from "../compile-gatsby-files"
import { siteMetadata } from "./fixtures/utils/site-metadata"
import { moreDataConfig } from "./fixtures/utils/more-data-config"
import { readFile, remove, pathExists } from "fs-extra"

const dir = {
  js: `${__dirname}/fixtures/js`,
  ts: `${__dirname}/fixtures/ts`,
}

jest.mock(`@parcel/core`, () => {
  const parcelCore = jest.requireActual(`@parcel/core`)

  const { Parcel: OriginalParcel } = parcelCore

  class MockedParcel extends OriginalParcel {
    constructor(options) {
      super(options)
      this.options = options // Expose for assertions
    }
  }

  return {
    ...parcelCore,
    Parcel: MockedParcel,
  }
})

interface IMockedParcel extends Parcel {
  options: unknown
}

describe(`gatsby file compilation`, () => {
  describe(`constructBundler`, () => {
    it(`should construct Parcel relative to passed directory`, () => {
      const { options } = constructParcel(dir.js) as IMockedParcel

      expect(options).toMatchSnapshot({
        entries: [
          `${dir.js}/${gatsbyFileRegex}`,
          `${dir.js}/plugins/**/${gatsbyFileRegex}`,
        ],
        targets: {
          root: {
            distDir: `${dir.js}/${COMPILED_CACHE_DIR}`,
          },
        },
        cacheDir: `${dir.js}/${PARCEL_CACHE_DIR}`,
      })
    })
  })

  describe(`compileGatsbyFiles`, () => {
    afterEach(async () => {
      for (const directory of [dir.js, dir.ts]) {
        await remove(`${directory}/.cache`)
      }
    })

    it(`should not compile gatsby-config.js`, async () => {
      await compileGatsbyFiles(dir.js)

      const isCompiled = await pathExists(
        `${dir.js}/.cache/compiled/gatsby-config.js`
      )
      expect(isCompiled).toEqual(false)
    })

    it(`should compile gatsby-config.ts`, async () => {
      await compileGatsbyFiles(dir.ts)

      const compiledGatsbyConfig = await readFile(
        `${dir.ts}/.cache/compiled/gatsby-config.js`,
        `utf-8`
      )

      expect(compiledGatsbyConfig).toContain(siteMetadata.title)
      expect(compiledGatsbyConfig).toContain(siteMetadata.siteUrl)
      expect(compiledGatsbyConfig).toContain(moreDataConfig.options.name)
    })

    it(`should not compile gatsby-node.js`, async () => {
      await compileGatsbyFiles(dir.js)

      const isCompiled = await pathExists(
        `${dir.js}/.cache/compiled/gatsby-node.js`
      )
      expect(isCompiled).toEqual(false)
    })

    it(`should compile gatsby-node.ts`, async () => {
      await compileGatsbyFiles(dir.ts)

      const compiledGatsbyNode = await readFile(
        `${dir.ts}/.cache/compiled/gatsby-node.js`,
        `utf-8`
      )

      expect(compiledGatsbyNode).toContain(`I am working!`)
    })
  })
})
