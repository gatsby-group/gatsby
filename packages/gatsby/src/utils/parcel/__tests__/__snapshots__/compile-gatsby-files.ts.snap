// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby file compilation constructBundler should construct Parcel relative to passed directory 1`] = `
Object {
  "cacheDir": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/.cache/.parcel-cache",
  "defaultConfig": "gatsby-parcel-config",
  "entries": Array [
    "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/gatsby-+(node|config).{ts,js}",
    "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/plugins/**/gatsby-+(node|config).{ts,js}",
  ],
  "mode": "production",
  "targets": Object {
    "root": Object {
      "distDir": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/.cache/compiled",
      "engines": Object {
        "node": ">= 14.15.0",
      },
      "includeNodeModules": false,
      "outputFormat": "commonjs",
      "sourceMap": false,
    },
  },
}
`;
